crates/admin/src/cluster_controller/logs_controller.rs

```diff
@@ -304,49 +305,196 @@ fn build_new_replicated_loglet_configuration(
+        &replication,
+        PlacementStrategy::AllAvailable, // todo: make placement strategy configurable
+        previous_configuration
+            .map(|config| NodeSet::from_single(config.sequencer.as_plain()))
+            .unwrap_or_default(),
+        &mut rng,
+    );
+
+    match placement {
+        Ok(nodeset) => {
+            // If the previous sequencer is still alive, irrespective of its generation, keep it.
+            // This might just avoid an unnecessary partition processor leadership change, even if the node generation is newer.
+            let sequencer = *previous_configuration
+                .and_then(|config| {
+                    observed_cluster_state
+                        .alive_nodes
+                        .get(&config.sequencer.as_plain())
+                })
+                .unwrap_or_else(|| {
+                    // If the previous sequencer is dead, pick a new one from the nodeset
+                    let new_sequencer = nodeset
+                        .iter()
+                        .choose(&mut rng)
+                        .expect("one node must exist");
+
+                    observed_cluster_state
+                        .alive_nodes
+                        .get(new_sequencer)
+                        .expect("node must be alive")
+                });
+
+            Some(ReplicatedLogletParams {
+                loglet_id,
+                sequencer,
+                replication,
+                nodeset,
+                write_set: None,
+            })
+        }
+        Err(PlacementError::InsufficientWriteableNodes) => {
+            debug!(
+                ?loglet_id,
+                "Insufficient writeable nodes to place replicated loglet"
+            );
+            None
+        }
+    }
+}
+
+#[derive(
+    Debug, Clone, Eq, PartialEq, derive_more::Display, derive_more::Into, derive_more::IntoIterator,
+)]
+struct WritableNodeSet(NodeSet);
+
+impl WritableNodeSet {
+    /// Constructs a new nodeset consisting of only alive, read-write storage nodes.
+    pub fn from(cluster_state: &ObservedClusterState, nodes_config: &NodesConfiguration) -> Self {
+        Self(
+            cluster_state
+                .alive_nodes
+                .keys()
+                .copied()
+                .filter(|node_id| {
+                    nodes_config
+                        .get_log_server_storage_state(node_id)
+                        .can_write_to()
+                })
+                .collect(),
+        )
+    }
+
+    pub fn len(&self) -> usize {
+        self.0.len()
+    }
+}
+
+/// Placement selector for picking a set of storage nodes for a replicated loglet out of a
+/// broader pool of available nodes.
+///
+/// This selector can be reused once constructed to make multiple placements in a single scheduling iteration
+/// if the node configuration and the replication settings are not changing.
+#[cfg(feature = "replicated-loglet")]
```

Comment:
Is it feature gated because it is only used from code from this feature?

Reply:
Yeah, exactly, just to suppress the unused code warning.

Reply:
I inlined this as it wasn't adding much value.